password/src/Data/Password/Validate.hs

```diff
@@ -0,0 +1,253 @@
+{-# LANGUAGE CPP               #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+
+{-|
+Module      : Data.Password.Valid
+Copyright   : (c) Dennis Gosnell, 2019; Felix Paulusma, 2020; Hiroto Shioi, 2020
+License     : BSD-style (see LICENSE file)
+Maintainer  : cdep.illabout@gmail.com
+Stability   : experimental
+Portability : POSIX
+
+= validate
+
+It is common for passwords to have a set of requirements.
+(e.g. Password must contain at least eight characters that consist of alphabetic
+characters combined with numbers or special characters.)
+
+Validate module provides set of functions which enables you to validate them.
+
+-}
+
+module Data.Password.Validate
+  ( -- * Data type
+    PasswordPolicy (..),
+    InvalidReason (..),
+    CharacterCategory(..),
+    -- * Rules
+    CharSetPredicate(..),
+    mkCharSetPredicate,
+    -- * Default
+    defaultPasswordPolicy,
+    defaultCharSetPredicate,
+    -- * Functions
+    isValidPassword,
+    validatePassword,
+    -- * Utility
+    isValidPasswordPolicy,
+    specialCharacterSet,
+    defaultCharacterSet,
+  ) where
+
+import Data.Char (isAsciiLower, isAsciiUpper, isDigit)
+import Data.Maybe (catMaybes, fromMaybe)
+import Data.Password.Internal (Password (..))
+#if! MIN_VERSION_base(4,13,0)
+import Data.Semigroup ((<>))
+#endif
+import Data.Text (Text)
+import qualified Data.Text as T
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+--
+-- Import needed libraries.
+--
+-- >>> import Data.Password
+
+-- | Set of policies used to validate 'Password'
+data PasswordPolicy = PasswordPolicy
+    { minimumLength  :: !Int
+    -- ^ Required password minimum length
+    , maximumLength  :: !Int
+    -- ^ Required password maximum length
+    , uppercaseChars :: !(Maybe Int)
+    -- ^ Required number of upper-case characters
+    , lowercaseChars :: !(Maybe Int)
+    -- ^ Required number of lower-case characters
+    , specialChars   :: !(Maybe Int)
+    -- ^ Required number of special characters
+    , digitChars     :: !(Maybe Int)
+    -- ^ Required number of ASCII-digit characters
+    } deriving (Eq, Ord, Show)
+
+-- | Default value for the 'PasswordPolicy'
+defaultPasswordPolicy :: PasswordPolicy
+defaultPasswordPolicy = PasswordPolicy
+  { minimumLength = 8,
+    maximumLength = 32,
+    uppercaseChars = Just 1,
+    lowercaseChars = Just 1,
+    specialChars = Nothing,
+    digitChars = Just 1
+  }
+
+-- | Predicate which defines the characters that can be used for a password.
+newtype CharSetPredicate = CharSetPredicate
+  { getCharSetPredicate :: Char -> Bool
+  }
```

Comment:
Introduce `CharSetPredicate` type which is used to define characters that can be used for a password.